File: src/Web/Dart/decorator.dart
typedef Render = String Function();

String plain() {
  return 'alert';
}

Render audit(Render inner) {
  String audited() {
    return 'audit(${inner()})';
  }

  return audited;
}

Render encrypt(Render inner) {
  String encrypted() {
    return 'enc(${inner()})';
  }

  return encrypted;
}

void main() {
  final Render base = plain;
  final audited = audit(base);
  final encrypted = encrypt(base);
  final stacked = audit(encrypt(base));

  print('base=${base()}');
  print('audit=${audited()}');
  print('encrypted=${encrypted()}');
  print('stacked=${stacked()}');
}

File: src/Niche/Nim/decorator.nim
type Render = proc (): string {.closure.}

proc plain(): string = "alert"

proc audit(inner: Render): Render =
  result = proc (): string = "audit(" & inner() & ")"

proc encrypt(inner: Render): Render =
  result = proc (): string = "enc(" & inner() & ")"

let base: Render = plain
let audited = audit(base)
let encrypted = encrypt(base)
let stacked = audit(encrypt(base))

echo "base=" & base()
echo "audit=" & audited()
echo "encrypted=" & encrypted()
echo "stacked=" & stacked()
